can/io/logger.py

```diff
@@ -28,36 +29,42 @@ class Logger(BaseIOHandler, Listener):  # pylint: disable=abstract-method
       * .csv: :class:`can.CSVWriter`
       * .db: :class:`can.SqliteWriter`
       * .log :class:`can.CanutilsLogWriter`
-      * other: :class:`can.Printer`
+      * .txt :class:`can.Printer`
+
+    The **filename** may also be *None*, to fall back to :class:`can.Printer`.
 
     The log files may be incomplete until `stop()` is called due to buffering.
 
     .. note::
-        This class itself is just a dispatcher, and any positional an keyword
+        This class itself is just a dispatcher, and any positional and keyword
         arguments are passed on to the returned instance.
     """
 
+    # pylint: disable=too-many-return-statements
     @staticmethod
-    def __new__(cls, filename, *args, **kwargs):
+    def __new__(
+        cls, filename: typing.Optional[can.typechecking.PathLike], *args, **kwargs
+    ):
         """
-        :type filename: str or None or path-like
-        :param filename: the filename/path the file to write to,
-                         may be a path-like object if the target logger supports
-                         it, and may be None to instantiate a :class:`~can.Printer`
-
+        :param filename: the filename/path of the file to write to,
+                         may be a path-like object or None to
+                         instantiate a :class:`~can.Printer`
         """
-        if filename:
-            if filename.endswith(".asc"):
-                return ASCWriter(filename, *args, **kwargs)
-            elif filename.endswith(".blf"):
-                return BLFWriter(filename, *args, **kwargs)
-            elif filename.endswith(".csv"):
-                return CSVWriter(filename, *args, **kwargs)
-            elif filename.endswith(".db"):
-                return SqliteWriter(filename, *args, **kwargs)
-            elif filename.endswith(".log"):
-                return CanutilsLogWriter(filename, *args, **kwargs)
-
-        # else:
-        log.warning('unknown file type "%s", falling pack to can.Printer', filename)
-        return Printer(filename, *args, **kwargs)
+        if filename is None:
+            return Printer(*args, **kwargs)
+
+        suffix = pathlib.PurePath(filename).suffix
+        if suffix == ".asc":
```

Comment:
You can switch this to a table-driven method, which I think is cleaner and allows you to skip disabling the linter. Something like:

```python
        lookup = {
            ".asc": ASCWriter,
            ".blf": BLFWriter,
            ".csv": CSVWriter,
            ".db": SqliteWriter,
            ".log": CanutilsLogWriter,
            ".txt": Printer,
        }

        try:
            return lookup[suffix](filename, *args, **kwargs)
        except KeyError:
            pass

```

Reply:
Sounds good, I'll have a look into this. :smile: